### PR TITLE
test: add a bunch of ASSERT_INT macros

### PR DESCRIPTION
**Pull request by santigimeno**
To make the debugging of test issues easier.

The following integer macros are added:

`ASSERT_EQ(a, b)`, `ASSERT_GE(a, b)`, `ASSERT_GT(a, b)`,
`ASSERT_LE(a, b)`, `ASSERT_LT(a, b)` and `ASSERT_NE(a, b)`.

And its corresponding unsigned integer macros:

`ASSERT_UINT64_EQ(a, b)`, `ASSERT_UINT64_GE(a, b)`,
`ASSERT_UINT64_GT(a, b)`, `ASSERT_UINT64_LE(a, b)`,
`ASSERT_UINT64_LT(a, b)` and `ASSERT_UINT64_NE(a, b)`.

Also these macros for `NULL` and pointer checks:

`ASSERT_NULL(a)`, `ASSERT_NOT_NULL(a)`, `ASSERT_PTR_EQ(a, b)` and
`ASSERT_PTR_NE(a, b)`.

And finally these macros for strings and buffers:

`ASSERT_STR_EQ(a, b)`/`ASSERT_STR_NEQ(a, b)` that use the `strcmp()`
call.
`ASSERT_MEM_EQ(a, b)`/`ASSERT_MEM_NEQ(a, b)` and
`ASSERT_MEM_HEX_EQ(a, b)`/`ASSERT_MEM_HEX_NEQ(a, b)` that use the
`memcmp()` call. The former, prints the data in string format and the
latter in hex format.

These macros are used in the following way:
```c
ASSERT_EQ(UV_EINVAL, uv_loop_close(&loop));
```

With a sample output that would be as follows:
```
Assertion failed in test/test-loop-close.c on line 44: `UV_EINVAL == uv_loop_close(&loop)` (-22 == -16)
```

To view multiples examples if their use, the `test-ipc.c` file has been
modified to use these macros.

The `_ISOC99_SOURCE` is defined to support `inttypes.h` in `z/OS`.

### CODE REVIEW

**Comment by bnoordhuis** [comment]
Seems like a good idea to me. Node/V8/Chromium have a whole family of such macros: ASSERT_EQ/ASSERT_NE/ASSERT_LT/etc.

**Comment by santigimeno** [comment]
Updated to support multiple `ASSERT_INT` macros.

**Comment by santigimeno** [comment]
Added macros for strings and buffers.

**Comment by santigimeno** [comment]
Renamed macros to be more aligned with what Ben suggested:
`ASSERT_INT_EQ`, `ASSERT_INT_GE`, `ASSERT_INT_GT`, `ASSERT_INT_LE`, `ASSERT_INT_LT`, `ASSERT_INT_NE`, etc.
Also added the macros to `test-ipc.c` to show how it would work and maybe catch what's going on with `ipc_closed_handle`.

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1808/

File: test/task.h [hunk 1]
```diff
@@ -109,6 +109,134 @@ typedef enum {
...
+#define ASSERT_LONG_LONG_INT_LE(a, b) ASSERT_INT_BASE(a, <=, b, long long int, lli)
+#define ASSERT_LONG_LONG_INT_LT(a, b) ASSERT_INT_BASE(a, <, b, long long int, lli)
+#define ASSERT_LONG_LONG_INT_NE(a, b) ASSERT_INT_BASE(a, !=, b, long long int, lli)
+
+#define ASSERT_LONG_LONG_UINT_EQ(a, b) ASSERT_INT_BASE(a, ==, b, unsigned long long int, llu)
+#define ASSERT_LONG_LONG_UINT_GE(a, b) ASSERT_INT_BASE(a, >=, b, unsigned long long int, llu)
+#define ASSERT_LONG_LONG_UINT_GT(a, b) ASSERT_INT_BASE(a, >, b, unsigned long long int, llu)
+#define ASSERT_LONG_LONG_UINT_LE(a, b) ASSERT_INT_BASE(a, <=, b, unsigned long long int, llu)
+#define ASSERT_LONG_LONG_UINT_LT(a, b) ASSERT_INT_BASE(a, <, b, unsigned long long int, llu)
+#define ASSERT_LONG_LONG_UINT_NE(a, b) ASSERT_INT_BASE(a, !=, b, unsigned long long int, llu)
```
Comment (bnoordhuis) [inline test/task.h]: This can probably be simplified a little. It's safe to widen any signed or unsigned integral type to `int64_t` (except `uint64_t`) so only two sets of macros are needed.

For the sake of brevity, I'd call the first one `ASSERT_EQ(...)`, etc., and the second one `ASSERT_UINT64_EQ(...)`.

File: test/task.h [hunk 2]
```diff
@@ -109,6 +109,134 @@ typedef enum {
...
+
+#define ASSERT_STR_EQ(a, b) \
+  ASSERT_BASE(strcmp(a, b) == 0, a, ==, b, s)
+
+#define ASSERT_STR_NE(a, b) \
+  ASSERT_BASE(strcmp(a, b) != 0, a, !=, b, s)
+
+#define ASSERT_MEM_EQ(a, b, size) \
+  ASSERT_BASE_LEN(memcmp(a, b, size) == 0, a, ==, b, s, size)
```
Comment (bnoordhuis) [inline test/task.h]: Wouldn't you always want to dump the contents of `a` and `b` on inequality?

Comment (bnoordhuis) [reply test/task.h]: Let me rephrase: `memcmp()` implies we're comparing binary data, not strings, so does it make sense to try to print them as strings?

(I realize that most of the time we're comparing buffers containing ASCII text so the answer may well be "yes, it does.")

Perhaps a middle ground is an `xxd()` like function that dumps hex and text? I'm sure I have one in one of my projects that I could dig up. Doesn't have to hold up this PR though.

File: test/test-ipc.c [hunk 3]
```diff
@@ -247,37 +248,39 @@ static void on_read_listen_after_bound_twice(uv_stream_t* handle,
 
   fprintf(stderr, "got %d bytes\n", (int)nread);
 
-  ASSERT(uv_pipe_pending_count(pipe) > 0);
+  ASSERT_GT(uv_pipe_pending_count(pipe), 0);
   pending = uv_pipe_pending_type(pipe);
-  ASSERT(nread > 0 && buf->base && pending != UV_UNKNOWN_HANDLE);
+  ASSERT_GT(nread, 0);
+  ASSERT(buf->base);
```
Comment (bnoordhuis) [inline test/test-ipc.c]: FWIW, Node has `ASSERT_NULL` and `ASSERT_NOT_NULL` macros. I suppose `ASSERT_PTR_EQ` and `ASSERT_PTR_NE` macros might come in handy too.

File: test/task.h (hunk 2 above)
Comment (santigimeno) [reply test/task.h]: That's the reason I created the `HEX` versions for the case of binary data. Anyway, if you could share some code, it would be great.

File: test/test-ipc.c (hunk 3 above)
Comment (santigimeno) [reply test/test-ipc.c]: Yes, I thought about it. I'll add them then.

**Comment by santigimeno** [comment]
Added NULL and PTR macros and also a fix for the STR macros.

File: test/task.h [hunk 4]
```diff
@@ -193,10 +209,10 @@ typedef enum {
 #define ASSERT_UINT64_NE(a, b) ASSERT_INT_BASE(a, !=, b, uint64_t, PRIu64)
 
 #define ASSERT_STR_EQ(a, b) \
-  ASSERT_BASE(strcmp(a, b) == 0, a, ==, b, s)
+  ASSERT_BASE(strcmp(a, b) == 0, a, ==, b, char*, s)
```
Comment (bnoordhuis) [inline test/task.h]: Just pass `"s"` (and `"p"`) to `ASSERT_BASE? No need for a separate ASSERT_INT_BASE macro.

Comment (santigimeno) [reply test/task.h]: Yeah, much nicer now. Thanks!

**Comment by santigimeno** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1816/

**Comment by richardlau** [comment]
It looks like this is failing to compile on z/OS (I've just swapped the z/OS builds to use cmake (https://github.com/libuv/libuv/pull/2737#issuecomment-603276474) but the above CI run fails because it doesn't include https://github.com/libuv/libuv/pull/2737 so I ran the gyp based build manually):
https://ci.nodejs.org/job/libuv-test-commit-zos/2333/nodes=zos-s390x/console
```
   /bin/xlc '-D_LARGEFILE_SOURCE' '-D_FILE_OFFSET_BITS=64' '-D_UNIX03_THREADS' '-D_UNIX03_SOURCE' '-D_UNIX03_WITHDRAWN' '-D_OPEN_SYS_IF_EXT' '-D_OPEN_SYS_SOCK_EXT3' '-D_OPEN_SYS_SOCK_IPV6' '-D_OPEN_MSGQ_EXT' '-D_XOPEN_SOURCE_EXTENDED' '-D_ALL_SOURCE' '-D_LARGE_TIME_API' '-D_OPEN_SYS_FILE_EXT' '-D_AE_BIMODAL' '-DPATH_MAX=255' '-DNDEBUG' -I../include  -qxplink -O3  -qmakedep=gcc -MF /u/iojs/build/workspace/libuv-test-commit-zos/nodes/zos-s390x/out/Release/.deps//u/iojs/build/workspace/libuv-test-commit-zos/nodes/zos-s390x/out/Release/obj.target/run-tests/test/test-ipc.o.d.raw  -q64 -c -o /u/iojs/build/workspace/libuv-test-commit-zos/nodes/zos-s390x/out/Release/obj.target/run-tests/test/test-ipc.o ../test/test-ipc.c
ERROR CCN3276 ../test/test-ipc.c:82    Syntax error: possible missing ')'?
ERROR CCN3275 ../test/test-ipc.c:82    Unexpected text ')' encountered.
ERROR CCN3276 ../test/test-ipc.c:88    Syntax error: possible missing ')'?
ERROR CCN3275 ../test/test-ipc.c:88    Unexpected text ')' encountered.
ERROR CCN3276 ../test/test-ipc.c:91    Syntax error: possible missing ')'?
ERROR CCN3275 ../test/test-ipc.c:91    Unexpected text ')' encountered.
ERROR CCN3276 ../test/test-ipc.c:105   Syntax error: possible missing ')'?
ERROR CCN3275 ../test/test-ipc.c:105   Unexpected text ')' encountered.
ERROR CCN3276 ../test/test-ipc.c:106   Syntax error: possible missing ')'?
ERROR CCN3275 ../test/test-ipc.c:106   Unexpected text ')' encountered.
ERROR CCN3276 ../test/test-ipc.c:140   Syntax error: possible missing ')'?
ERROR CCN3275 ../test/test-ipc.c:140   Unexpected text ')' encountered.
ERROR CCN3276 ../test/test-ipc.c:141   Syntax error: possible missing ')'?
CCN0793(I) Compilation failed for file ../test/test-ipc.c.  Object file not created.
```

**Comment by richardlau** [comment]
For z/OS I think we'll need to define `_ISOC99_SOURCE` before including `inttypes.h`:
https://www.ibm.com/support/knowledgecenter/SSLTBW_2.1.0/com.ibm.zos.v2r1.bpxbd00/inttyph.htm

Something like (the CMakeLists.txt change made on top of current `v1.x`):
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index bacc92c..1444cdc 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -227,6 +227,7 @@ if(CMAKE_SYSTEM_NAME STREQUAL "OS390")
   list(APPEND uv_defines PATH_MAX=255)
   list(APPEND uv_defines _AE_BIMODAL)
   list(APPEND uv_defines _ALL_SOURCE)
+  list(APPEND uv_defines _ISOC99_SOURCE)
   list(APPEND uv_defines _LARGE_TIME_API)
   list(APPEND uv_defines _OPEN_MSGQ_EXT)
   list(APPEND uv_defines _OPEN_SYS_FILE_EXT)
diff --git a/uv.gyp b/uv.gyp
index 480cf6e..189a89c 100644
--- a/uv.gyp
+++ b/uv.gyp
@@ -16,6 +16,7 @@
       }],
       ['OS=="zos"', {
         'shared_zos_defines': [
+          '_ISOC99_SOURCE',
           '_UNIX03_THREADS',
           '_UNIX03_SOURCE',
           '_UNIX03_WITHDRAWN',
```

**Comment by santigimeno** [comment]
CI looks good now after adding the `z/OS` changes: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1823/

**Comment by santigimeno** [comment]
Landed in https://github.com/libuv/libuv/commit/53caf044e39476c4888f21a79e425391027cdd58. Thanks!
